src/lib.rs

```diff
@@ -462,49 +473,70 @@ pub struct DataFormatDescriptor<'data> {
     pub data: &'data [u8],
 }
 
-pub struct BasicDataFormatDescriptor<'data> {
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct BasicDataFormatDescriptorHeader {
     /// None means Unspecified or is an otherwise unknown value
     pub color_model: Option<ColorModel>, //: 8;
     /// None means Unspecified or is an otherwise unknown value
     pub color_primaries: Option<ColorPrimaries>, //: 8;
     /// None means Unspecified or is an otherwise unknown value
     pub transfer_function: Option<TransferFunction>, //: 8;
-    pub flags: DataFormatFlags,           //: 8;
-    pub texel_block_dimensions: [u32; 4], //: 8 x 4;
-    pub bytes_planes: [u32; 8],           //: 8 x 8;
-    sample_data: &'data [u8],
+    pub flags: DataFormatFlags,          //: 8;
+    pub texel_block_dimensions: [u8; 4], //: 8 x 4;
+    pub bytes_planes: [u8; 8],           //: 8 x 8;
 }
 
-impl<'data> BasicDataFormatDescriptor<'data> {
-    pub fn parse(bytes: &'data [u8]) -> Result<Self, ParseError> {
+impl BasicDataFormatDescriptorHeader {
+    pub const LENGTH: usize = 16;
+
+    pub fn as_bytes(&self) -> [u8; Self::LENGTH] {
+        let mut bytes = [0u8; Self::LENGTH];
+
+        let color_model = self.color_model.map(|c| c.value()).unwrap_or(0);
+        let color_primaries = self.color_primaries.map(|c| c.value()).unwrap_or(0);
+        let transfer_function = self.transfer_function.map(|t| t.value()).unwrap_or(0);
+
+        let texel_block_dimensions = self.texel_block_dimensions.map(|dim| dim - 1);
+
+        bytes[0] = color_model;
+        bytes[1] = color_primaries;
+        bytes[2] = transfer_function;
+        bytes[3] = self.flags.bits;
+        bytes[4..8].copy_from_slice(&texel_block_dimensions);
+        bytes[8..16].copy_from_slice(&self.bytes_planes);
+
+        bytes
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ParseError> {
         let mut offset = 0;
 
         let v = bytes_to_u32(bytes, &mut offset)?;
-        let model = shift_and_mask_lower(0, 8, v);
-        let primaries = shift_and_mask_lower(8, 8, v);
-        let transfer = shift_and_mask_lower(16, 8, v);
-        let flags = shift_and_mask_lower(24, 8, v);
+        let model = shift_and_mask_lower(0, 8, v) as u8;
```

Comment:
I think I'd like to make another variant for this function or make it generic to reduce the noise in all these cases.

Reply:
Generic over simple bounds seems reasonable.